## Exploring eBay Car Sales Data
In this project we are are going to work on the dataset of used cars from _ebay_. The dataset was originally a part of kaggle but can no longer be found there, but can still access it through following [link](https://data.world/data-society/used-cars-data)

In this project we are going to work on simple tasks of data cleaning and basic descriptive analysis to get some meaningful insight from the data.

In [1]:
# let's import required libraries that we'll be using in this project
import numpy as np
import pandas as pd

pd.set_option('display.max_columns', None)

In [2]:
# loading our dataset from csv file
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Right from the bat we can observe that there are 50,000 rows/entries and 20 columns to define these records.

Column name has to be snake-casing instaed of camel-casing.  

We can observe that certain columns/fields(vehicleType, gearbox, model, fuelType, notRepairedDamage) in dataset have missing values. We'll need to figure out what to do with this missing values when we dive deeper into analysis. 

Also, we can observe that data types of certain columns are in improper format which makes analysis difficult. We'll need to fix this data types to get reliable insights on data.

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

After having a closer look at the data for the first few rows we can see that, data type date columns and certain numerical columns are in object type which has to be corrected

### Cleaning Column Names
* As observed above column names are in Camel-casing format, in following cells we'll perform the task of converting these names to snake-casing format

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [8]:
autos.columns = new_columns

In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In the above cells we modified the camel-casing format of column names to desired and standarized snake-casing format. This is the preferred format of column names with which most data scientist/analyst to work with for analysis and modeling 

### Initial Exploration and Cleaning
- Here we'll have a closer look at values present in each column. Qualitative data columns/text columns where almost all values are same can often be dropped as being constant they don't add any meaningful information to the analysis
- We'll also perform data type transformation on certain columns so that it appears in it's proper format for carrying out analysis. Also, if required we'll change their respective column names in order to maintain transparency throughout the team regarding their units.

In [10]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Having a look at brief summary of the columns we can observe that:
- `seller`, `offer_type` does add much value to analysis. Since out of 50,000 records 49,999 records represents the same value for them. It acts as constant values in our analysis and better if these columns are dropped off.
- `nr_of_pictures` column/field can also be dropped off since there's just a single value present in every record for that column. Again it acts as a constant and doesn't add much value to the analysis

Now, we are going perform some data transformtion on columns where data is suppose to be in numerical format i.e. `price` and `odometer` columns/variables. These are of object type because of presence an additional caharacter in them which doesn't fit in numerical type. 

Steps for converting type:
- first we'll replace non-numerical characters with empty string('') using Series.str.replace() method
- We'll then perform type conversion using Series.astype() method

In [11]:
# droping columns that won't be useful in analysis
col_to_drop = ['seller','offer_type', 'nr_of_pictures']
autos.drop(col_to_drop, axis=1, inplace=True)

In [12]:
# checking the format in which price in represnted in price column
autos.price.unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

We can see that '$' and ',' act as special characters in the string and need to be eliminated before converting to numerical data type

In [13]:
autos['price'] = autos.price.str.replace('$','').str.replace(',','').astype(int)

In [14]:
# checking for format and special cases in odometer column
autos.odometer.unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

Again, we observe that 'km' and ',' has to be eliminated form the values present in odometer column before converting it numerical type 

After converting the data type to numerical, we'll also make changes to the column name just to make it clear for other team member about the data present in the column for futher transparacy in communication

In [15]:
autos['odometer'] = autos.odometer.str.replace('km','')\
                    .str.replace(',', '').astype(int)
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

In [16]:
autos.dtypes

date_crawled          object
name                  object
price                  int64
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer_km            int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
postal_code            int64
last_seen             object
dtype: object

### Exploring the Odometer and Price Columns
- In following cells we're going to explore `odometer` and `price` column even further to look for probable outlier in the data. 
- Outliers can often result in false interpretation of data and can result in bringing out misleading insights to stakeholders

In [17]:
# checking number of unique values in the column
autos.price.nunique()

2357

In [18]:
# using Series.decribe() method to get a basic summary stat about the column
autos.price.describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [19]:
autos.price.value_counts(dropna=False).sort_index(ascending=True).head(15)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price, dtype: int64

In [20]:
autos.price.value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

- There seems to be a huge jump in values from \\$350,000 onwards. Any value after that will skew the data towards right, so it's better to eliminate values greater than \\$350,000
- Also, reselling a car for free or \\$0 seems a bit off in the data. So, we are going to eliminate those values too considering them to be outliers

In [21]:
# Filter data to eliminate extreme values from the data
autos = autos[autos.price.between(1,350000)]

In [22]:
autos.odometer_km.nunique()

13

In [23]:
autos.odometer_km.describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [24]:
autos.odometer_km.value_counts().sort_index()

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

It seems there are no extreme outliers in the data. Having a look at data, it seems to be left-skewed, but the minimum value seems to be legitimate value and no need to drop it form the dataset at the moment. 

### Exploring the date columns

In [25]:
print(autos[['date_crawled', 'ad_created','last_seen']].dtypes)
autos[['date_crawled', 'ad_created','last_seen']][0:5]

date_crawled    object
ad_created      object
last_seen       object
dtype: object


date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [26]:
(autos['date_crawled']
         .str[:10]
         .value_counts(normalize=True, dropna=True)
         .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [27]:
(autos['date_crawled']
         .str[:10]
         .value_counts(normalize=True, dropna=True)
         .sort_values()
        )

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

From above observations we can see that the site was scraped on daily basis for a period of roughly about a month(between 5th, March 2016 to 7th, April 2016) 

In [28]:
(autos['last_seen']
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

We observe from the last 3 days that there was a massive spike in last-seen values. Last-seen values indicates the day on which the car was last seen on website by the crawler, probable because it got sold.\
The reason for spike in last 3 days is highly unlikely due to hike in sales because rest of the days seems pretty consistent in terms of sales. This spike/disproportionality  in last-seen column can be due to end of crawling period.

In [29]:
(autos['ad_created']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index()
        )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

This column indicates the date on which an ad was created for a particular car. While most car's ad created date fall with 1-2 month of listing date, there are certain cars who's ad were quite old, with oldest been around 9 months

In [30]:
autos.registration_year.describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Registration year can turn out to be a good indicator to find out how old the car is. There's one odd thing which we can observe form the summary statistics, the minimum registration year is 1000(indicating first car registered in year 1000) even before when cars were invented and maximum year indicate 9999 which is in the future. 

### Dealing with Incorrect Registration Year Data

In [31]:
autos.registration_year.value_counts().sort_index().head(10)

1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
Name: registration_year, dtype: int64

In [32]:
autos.registration_year.value_counts().sort_index(ascending=False).head(15)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
Name: registration_year, dtype: int64

In [33]:
threshold = 5*autos.shape[0]/100
false_reg_year_count = (~(autos['registration_year'].between(1910,2016))).sum()
print(false_reg_year_count)
if threshold > false_reg_year_count:
    print("We can drop incorrect records")
    print(false_reg_year_count*100/autos.shape[0])

1884
We can drop incorrect records
3.8793369710697005


From above observations and having details about when the data was been crawled, we can conclude that our value must be between 1910 and 2016. Since cars were invented after 1900, and data was crawled in 2016 our range of values has to be between this years anything outside this range are false entries

As a general rule in analytics if total missing values/inaccurate entries account for less than 5% of total data and eleimination of those records won't hinder our analysis, then it's a good practice to eliminate those rows to carry out more reliable analysis

Here these false entries account for around 4% of total data so we can eliminate those records form the dataset. 

In [34]:
autos = autos[autos['registration_year'].between(1910,2016)]

In [35]:
(autos.registration_year
    .value_counts(normalize=True)
    .sort_index(ascending=True)
    .head(10)
    )

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
Name: registration_year, dtype: float64

In [36]:
(autos.registration_year
    .value_counts(normalize=True)
    .sort_index(ascending=False)
    .head(10)
    )

2016    0.026135
2015    0.008397
2014    0.014203
2013    0.017202
2012    0.028063
2011    0.034768
2010    0.034040
2009    0.044665
2008    0.047450
2007    0.048778
Name: registration_year, dtype: float64

It appears that most cars were registered in recent year rather than in earlier years, indicating that people in recent years tend to buy more cars than they used to do back in earlier years. There can be many reasons for this to occur, one possible reason can be cars were viewed as luxury and was not not much of a necessity back in the days. 

### Exploring Price by Brand
- In the follwoing section we are going to perform aggregation, but we'll be perform manual aggraegation without using and inbuilt pandas function like df.groupby() or df.pivot_table()

Steps for performing aggregation:
- Identify unique values of columns we wanted to aggregate by(aggregating column)
- create an empty dictionary
- loop over the unique value and inside loop perform the following task:
1. subset df for each unique values, perform our desired aggregation on required column on which we want to aggregate on(aggregated column)
2. assign result to dictionary with as the unique value of aggregating column as key  and aggregated values as its respective value

In [37]:
print(autos.brand.nunique())
autos.brand.value_counts(normalize=True)[:10]

40


volkswagen       0.211264
bmw              0.110045
opel             0.107581
mercedes_benz    0.096463
audi             0.086566
ford             0.069900
renault          0.047150
peugeot          0.029841
fiat             0.025642
seat             0.018273
Name: brand, dtype: float64

Observations:

- We can see that there are 40 unique brand of cars that are present on the website. For simplicity, we can see above distribution of top 10 most appearing brands. 
- We can see that most recurring brand is volkswagen comprising for about 21% of total cars appearing on site, followed by opel, bmw, mercedes benz, audi, ford each for which listing count on website is greater than 5%. 

For rest of the brands they don't appear so frequently it's less than 5%(So we won't be focusing on them)

In [39]:
percent_threshold = 0.05
top_brand = autos['brand'].value_counts(normalize=True)
top_brand = top_brand[top_brand>percent_threshold].index
top_brand_avg_price = {}
for ech_brand in top_brand:
    mean_price = round(autos.loc[autos['brand']==ech_brand,'price'].mean())
    top_brand_avg_price[ech_brand] = mean_price
    
top_brand_avg_price

{'volkswagen': 5402.0,
 'bmw': 8333.0,
 'opel': 2975.0,
 'mercedes_benz': 8628.0,
 'audi': 9337.0,
 'ford': 3749.0}

For the most appearing car brands on website we can make following observations for their price:
- Opel, Ford on average are the cheapest cars that are on the site
- BMW, mercedes Benz, Audi are the costliest once with their prices all ranging above 8300$ on average
- While, volkswagen has a sweet spot in between these price ranges at at 5402 average price which explain it's popularity 

### Storing aggregate data in DataFrame

In [41]:
top_brand_avg_mileage = {}

for ech_brand in top_brand:
    mean_mileage = round(autos.loc[autos['brand']==ech_brand,'odometer_km'].mean())
    top_brand_avg_mileage[ech_brand] = mean_mileage
    
top_brand_avg_mileage

{'volkswagen': 128707.0,
 'bmw': 132573.0,
 'opel': 129310.0,
 'mercedes_benz': 130788.0,
 'audi': 129157.0,
 'ford': 124266.0}

In [42]:
mean_price = pd.Series(top_brand_avg_price).sort_values(ascending=False)
mean_mileage = pd.Series(top_brand_avg_mileage)

In [46]:
brand_info = pd.DataFrame(mean_price, columns=['mean_price'])
brand_info['mean_mileage'] = mean_mileage

In [47]:
brand_info

mean_price  mean_mileage
audi               9337.0      129157.0
mercedes_benz      8628.0      130788.0
bmw                8333.0      132573.0
volkswagen         5402.0      128707.0
ford               3749.0      124266.0
opel               2975.0      129310.0

We don't see much variations in average mileage as we in average price